### DISCUSSION TITLE
/orm/relationships.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#350] [Announcements]
# /orm/relationships.html

[[toc]] Introduction It's common for database tables to be interconnected. For instance, a blog post may have many comments, or an order may be linked to the user who placed it....

https://www.goravel.dev/orm/relationships.html

**Comment by ruoxiao-zh** [comment]
Why did I use many to many model association and bring out the data that was soft deleted?

**Reply by hwbrzzl** [reply]
Sorry, didn't get you.

**Reply by ruoxiao-zh** [reply]
When querying the roles of users in the 'uses',' roles', and 'user_roles' tables, the deleted roles were also found

**Reply by ruoxiao-zh** [reply]
Used soft deleted

**Reply by ruoxiao-zh** [reply]
Query is:
SELECT * FROM `users` WHERE `users`.`id` = '1' AND `users`.`deleted_at` IS NULL
SELECT * FROM `user_roles` WHERE `user_roles`.`user_id` IN (1,3,5,7)
SELECT * FROM `roles` WHERE `roles`.`id` = 1 AND `roles`.`deleted_at` IS NULL

Why is not:
SELECT * FROM `user_roles` WHERE `user_roles`.`user_id` IN (1,3,5,7) AND `user_roles`.`deleted_at` IS NULL

**Reply by hwbrzzl** [reply]
I guess `user_roles` doesn't enable the soft_delete function.

**Reply by ruoxiao-zh** [reply]
type UserRole struct {
  Model
  UserId int64 `json:"user_id" gorm:"column:user_id"`
  RoleId int64 `json:"role_id" gorm:"column:role_id"`
  Timestamps
  SoftDeletes
}

type Model struct {
  ID int64 `gorm:"column:id;primaryKey;autoIncrement;" json:"id,omitempty"`
}

type Timestamps struct {
  CreatedAt carbon.DateTime `gorm:"autoCreateTime;column:created_at;index" json:"created_at,omitempty"`
  UpdatedAt carbon.DateTime `gorm:"autoUpdateTime;column:updated_at;index" json:"updated_at,omitempty"`
}

type SoftDeletes struct {
  DeletedAt gorm.DeletedAt `gorm:"column:deleted_at;index;" json:"-"`
}

**Reply by ruoxiao-zh** [reply]
type User struct {
  Model
  Uid   string `json:"uid" gorm:"column:uid"`
  Name  string `json:"name" gorm:"column:name"`
  Mail  string `json:"mail" gorm:"column:mail"`
  Phone string `json:"phone" gorm:"column:phone"`
  Timestamps
  SoftDeletes
}

type UserWithRole struct {
  User
  Roles []Role `json:"roles" gorm:"many2many:user_roles;joinForeignKey:userId;joinReferences:roleId"`
}

func (m *User) GetAllWithRoles() (userRoles []UserWithRole, err error) {
  err = m.GetQueryModel().With("Roles").Order("id asc").Get(&userRoles)
  return
}

**Reply by hwbrzzl** [reply]
I think this is not the best place to discuss this. Could you create a PR at https://github.com/goravel/example? I can check it out.

**Comment by klphp** [comment]
我想仿写一个laravel的token管理表，表结构如下：
```
type PersonalAccessTokens struct {
	ID            string          `json:"id" gorm:"primaryKey"`
	TokenableType string          `json:"tokenable_type"`
	TokenableID   uint            `json:"tokenable_id"`
	Name          string          `json:"name"`
	Token         string          `json:"token"`
	Abilities     string          `json:"abilities"`
	LastUsedAt    carbon.DateTime `json:"last_used_at"`
	ExpiresAt     carbon.DateTime `json:"expires_at"`
	CreatedAt     carbon.DateTime `gorm:"autoCreateTime;column:created_at" json:"created_at"`
	UpdatedAt     carbon.DateTime `gorm:"autoUpdateTime;column:updated_at" json:"updated_at"`
}

```
现在我先创建了一个用户，由于是多态的对应关系。根据文档上面的介绍如下：
```
facades.Orm().Query().Select(orm.Associations).Create(&user)
但这个好像是创建user同时创建token
因我是先创建用户后创建token，怎么单独写入关联数据。
还是说我必须手动设置
user.PersonalAccessTokens.TokenableType和user.PersonalAccessTokens.TokenableID
后再
facades.Orm().Query().Select(orm.Associations).Create(&user.PersonalAccessTokens)
```

**Reply by klphp** [reply]
解决了 ```
facades.Orm().Query().Model(&user).Association(`PersonalAccessTokens`).Append(&user.PersonalAccessTokens) ```

**Comment by jxgame** [comment]
分页的时候，能查关联表吗？

facades.Orm().Query().Order("id desc").Association("Info").Paginate(page, pageSize, &users, &total)

报错：
facades.Orm().Query().Order("id desc").Association("Info").Paginate undefined

**Reply by jxgame** [reply]
使用 .With("name") 可解决，但是为什么文档里没有这个呀

**Reply by hwbrzzl** [reply]
`Association` 之后的方法是对关联模型的查询：https://www.goravel.dev/zh/orm/relationships.html#%E5%88%9B%E5%BB%BA-%E6%9B%B4%E6%96%B0%E5%85%B3%E8%81%94

`With` 方法是查询主模型同时加载关联模型：https://www.goravel.dev/zh/orm/relationships.html#%E9%A2%84%E5%8A%A0%E8%BD%BD

两者使用场景不同，注意区分。

**Comment by jxgame** [comment]
表A User phone 表B UserLog user_id

现在通过表B查询 电话号码=xxx的用户的所有记录，我使用如下方法

```
    var data models.UserLog
    query := facades.Orm().Query()
	keyword := ctx.Request().Input("search_keyword", "")
	if keyword != "" {
		query = query.With("User", func(query orm.Query) orm.Query {
			return query.Where("phone LIKE ?", "%"+keyword+"%")
		})
	} else {
		query = query.With("User")
	}

    _ := query.Paginate(page, pageSize, &data, &total)
```

在上述代码中，这行代码报错，如何解决？或者说，我这样的需求应该很平常的了，如何实现它呢？
```
		query.With("User", func(query orm.Query) orm.Query {
			return query.Where("phone LIKE ?", "%"+keyword+"%")
		})
```

**Reply by jxgame** [reply]
在追问一下，如果我只需要User 表里的 id,phone 不要其它的字段，如何做呢？

**Reply by jxgame** [reply]
上述报错这里 最开始它自动导入的是
"github.com/goravel/framework/mocks/database/orm"
我手动改成
"github.com/goravel/framework/database/orm"
又报错 orm.Query 没有 这个方法。

**Reply by jxgame** [reply]
不好意思，在文档中找到正确的导入类
"github.com/goravel/framework/contracts/database/orm"

此时报错没有了，但是结果不是预期。
如果要查询A的一个手机号，它仍然返回了B表的全部记录，只是将 关联的结果置为NULL了。

它的SQL是这样的
SELECT * FROM `users` WHERE phone LIKE '%138%' AND `users`.`id` = 1 AND `users`.`deleted_at` IS NULL

想要的是 只得到138的手机号在B的记录，不是所有的，仅将User关联置 null呀。

**Reply by jxgame** [reply]
问 AI 说让用 WhereHas ，但没有这个方法呀，Laravel 倒是有，总之是想要Laravel 的WhereHas来查询关联呀

**Reply by jxgame** [reply]
目前解决办法是 查询A 表满足条件的 pluck ，再 whereIn到 B表 。但这样不优雅呀

**Reply by hwbrzzl** [reply]
是的，目前还没有 WhereHas 方法，已经存在一个 issue：https://github.com/goravel/goravel/issues/373。With 方法的确无法实现你的需求，目前你的解决方案是可以的。

**Comment by jxgame** [comment]
请教一下，我有poster、comments 两张表。
poster 一对多 comments   comments中有poster_Id 
在poster中用 comments []*comments 这样可以建立一对多的关系。
现在的问题是，我在 comments 中另有一个字段 postered_id 这个id 也对应poster的ID
我想 通过 postered_id = poster.id查找 
 在comments表中如何定义呢

**Reply by hwbrzzl** [reply]
```
// Poster model 
Comments []*Comment 
OtherComments []*Comment `gorm:"foreignKey:PosteredID"`
```

**Reply by jxgame** [reply]
我只需要查询关表的特定字段，该如何做呢？

**Reply by hwbrzzl** [reply]
eg:

```
query.Query().With("Books", func(query contractsorm.Query) contractsorm.Query {
  return query.Where("name = ?", "with_book0").Select("id", "user_id", "name")
}).Find(&user1, user.ID)
```

**Reply by jxgame** [reply]
多谢，这样可以将其它不需要的字段置为空。
仍然有两个问题向您请教。

1、 User表 UserInfo 表 Posts表 它们之前关联关系是 userInfo 表的的user_id 关联user.id  posts 中的user_id 关联user.id
现在查询 posts 表 当我 with("user.info")  的时候，user表信息也会全出来。期望的是只查询 userinfo的内容。

2、User表 UserInfo 表 userinfo.user_id =user .id. 然后呢，我在usesInfo表里添加了一个关联 User *User.  即是希望查询 userInfo表的时候 直接。with("user") 查询出User表。
然后，查询是没有出现问题。一切正常。
当我又有一个表 posts 它 posts.id=user.id ，问题来了，当我用 	var data models.Posts
	_ = ctx.Request().Bind(&data) 
Bind 这一行会报错了 runtime: goroutine stack exceeds 1000000000-byte limit
如果我去掉 userinfo 中的 User *User 问题会消失。

感谢您的回复。

**Reply by hwbrzzl** [reply]
1. 可以试试添加一个 with("user", ...select()...)
2. 需要单步调试下，如果实在解决不了，可以向 goravel/example 提交一个PR，我试下

**Reply by jxgame** [reply]
您好，我不太会提交PR。第2点的具体情况是
User表 一对一关联 Userinfo表
Userinfo 表 一对一反向关联 User表
第三个表 Posts ，如果一对一关联了User表，那么在ctx.Request().Bind()的时候，报错

**Reply by jxgame** [reply]
报错是： runtime: goroutine stack exceeds 1000000000-byte limit

**Comment by damon35868** [comment]
我在模型中定义了关联模型，当我使用with的时候是正常带出来的，但如果我平时不想关联的时候不使用with查询使，那个对应的关联字段会变成null，那个字段能否直接不输出了？类似laravel的with那样

**Reply by hwbrzzl** [reply]
`直接不输出` 意思是？截图看下？

**Reply by damon35868** [reply]
你好，辛苦看看这个截图，看看能不能明白我的意思

使用with：facades.Orm().Query().With("User").Paginate(page, pageSize, &data.Tasks, &data.Total)
参考截图：https://drive.google.com/file/d/14007MUy057dHEjrFz9jCG_Uzw-MjBojU/view?usp=drive_link

不使用with：facades.Orm().Query().Paginate(page, pageSize, &data.Tasks, &data.Total)
参考截图：https://drive.google.com/file/d/1n-jkhn5V3bQpeEr-SotAO5CqOT_EEY04/view?usp=drive_link

我想的是没使用with的时候直接user这个字段就剔除

**Reply by hwbrzzl** [reply]
无权限查看文件，你可以直接将截图放到回复里

**Reply by damon35868** [reply]
你好，我放开权限了，这个回复框没有上传图片的位置，拖拽也不行

**Reply by hwbrzzl** [reply]
这个 user 字段的返回需要自己控制一下，例如在 response 返回时做个 transform 转换下数据。

**Reply by damon35868** [reply]
好的，我目前就是手动剔除的，只是想看看有没官方的，感谢，目前我用到的东西都具备了，准备开始生产使用，期待作者的继续更新和跟进laravel功能，谢谢🙏

**Reply by hwbrzzl** [reply]
欢迎 PR Issue

**Comment by ARSENAL91** [comment]
你好，为什么我的预加载，跨库的时候，会显示报错rror 1146 (42S02): Table 'financial_center.oa_organizations' doesn't exist。还是查原model的库？我在关联库已经设置了

func (r *OrganizationModel) Connection() string {
	return "ybs_mysql"
}	

query.With("Group").Paginate(page, pageSize, &TmpPgRelations, &total)

type TmpPgRelation struct {
	orm.Model
	TheMonth         string                                 `json:"the_month" gorm:"type:date;column:the_month"`
	EmployeesId      int64                                  `json:"employee_id" gorm:"column:employees_id"`
	DwOrganizationId int                                    `json:"dw_organization_id" gorm:"column:dw_organization_id"`
	CorpId           uint                                   `json:"corp_id" gorm:"column:corp_id"`
	GroupId          uint                                   `json:"group_id" gorm:"column:group_id"`
	Group            *models_organization.OrganizationModel `gorm:"foreignKey:ID;references:GroupId" json:"group"`
}

**Reply by hwbrzzl** [reply]
不支持跨库的
